index.js

```diff
@@ -799,9 +955,137 @@ studio.internal = (function(proto) {
+            return;
+          }
+          if (serviceMsg.kind === proto.ServiceMessageKind.eDisconnect ||
+              serviceMsg.kind === proto.ServiceMessageKind.eError) {
+            transport.onclose && transport.onclose({ code: 1000, reason: 'Service closed' });
+            serviceInstances.delete(instanceKey);
+            return;
+          }
+        }
+      });
+
+      appConnection.sendServiceMessage(serviceId, instanceId, proto.ServiceMessageKind.eConnect);
+
+      transport.send = function(bytes) {
+        appConnection.sendServiceMessage(serviceId, instanceId, proto.ServiceMessageKind.eData, bytes);
+      };
+
+      transport.close = function() {
+        appConnection.sendServiceMessage(serviceId, instanceId, proto.ServiceMessageKind.eDisconnect);
+        serviceInstances.delete(instanceKey);
+      };
+
+      return { transport: transport, instanceId: instanceId };
+    }
+
+    this.onServicesReceived = function(services, metadata) {
+      availableServices.clear();
+      var compatVersion = metadata.compatVersion || 2;
+
+      // Only primary connection creates proxy connections (prevents infinite loop)
+      if (!isPrimaryConnection) {
+        for (var i = 0; i < services.length; i++) {
+          availableServices.set(services[i].serviceId, services[i]);
+        }
+        return;
+      }
+
+      for (var i = 0; i < services.length; i++) {
+        var service = services[i];
+        availableServices.set(service.serviceId, service);
+
+        var isProxyService = service.type === 'websocketproxy' ||
+                             service.type === 'studioapi' ||
+                             service.type === 'logserver';
+
+        if (isProxyService && compatVersion >= 4) {
+          var result = makeServiceTransport(service.serviceId, notificationListener, autoConnect);
+          var proxyConnection = new obj.AppConnection(result.transport, notificationListener, autoConnect);
+          serviceConnections.set(service.serviceId, proxyConnection);
+
+          if (appConnection.onServiceConnectionEstablished) {
+            appConnection.onServiceConnectionEstablished(proxyConnection);
+          }
+        } else if (service.metadata && service.metadata.ip_address && service.metadata.port) {
+          var serviceUrl = service.metadata.ip_address + ':' + service.metadata.port;
+          appConnection.connectToService(service, serviceUrl, metadata);
+        }
+      }
+    };
+
+    this.onServiceMessage = function(serviceMessage, metadata) {
+      var instanceKey = serviceMessage.serviceId + ':' + (serviceMessage.instanceId || 0);
+      var handler = serviceInstances.get(instanceKey);
+
+      if (handler && handler.onMessage) {
+        handler.onMessage(serviceMessage, metadata);
+      }
+    };
+
+    this.connectToService = function(serviceInfo, serviceUrl, metadata) {
+      if (serviceInfo.type === 'studioapi' && !serviceConnections.has(serviceInfo.serviceId)) {
```

Comment:
`serviceInfo.type === 'studioapi'` - as mentioned in other comment, this specific type was moved under metadata. 

See:
* https://bitbucket.seaonics.com:448/projects/CDPES/repos/cdpmetadatalib/browse/CDPMeta/Type/services.h#18
* https://bitbucket.seaonics.com:448/projects/CDP/repos/cdp/browse/cdplogger/ModelSource/ServerRunner.cpp#202
* https://bitbucket.seaonics.com:448/projects/CDP/repos/cdp/browse/studioapiserver/Source/StudioAPIServer.cpp#322